### ISSUE TITLE
Multiple Suggestions: 1) Holding 2) Routes between fixes 3) Select all airplanes

### ISSUE THREAD

**Issue by nielsvdweide** [zlsa/atc#174]
Hi there! I have got a couple of suggestions to make this atc simulator a bit more real so let me know what you think.

1) Holding patterns. 
At this point, when you send an aircraft into a hold it will just circle either left or right. In commercial aviation a hold consists of a 180 degree turn, then 1 minute outbound, then again a 180 degree turn and another 1 minute leg. They make this 1 minute straight flying so the aircraft equipment will not suffer from conditions caused by being in a constant turn. Therefore, i would like to suggest to make the holding pattern look more real.

<img width="200" alt="schermafbeelding 2015-09-10 om 14 08 46" src="https://cloud.githubusercontent.com/assets/14002770/9787976/33e9aa5a-57c7-11e5-8a98-249991ce2ce1.png"> 

![holding_pattern-e1400680150992](https://cloud.githubusercontent.com/assets/14002770/9788008/70625d4c-57c7-11e5-81e1-c0ae2b51c1fe.jpg)

Eventually, we can also implement this holding technique to make aircraft hold over a fix instead of just circling in relation to nothing. For example: ACA627 f RTM then hold right. The aircraft will fly to rotterdam and then start a holding pattern over that fix. Ofcourse this is for later concern but then we also need to implement the correct entry of a holding. More later! 

2) Routes between fixes
Another thing i came across is when you make an airplane fly to a couple of fixes, including some of the with a big angle to turn. Normally, an airplane will fly as effective as possible, so they will turn a bit early of the fix to get on a straight line between the two fixes. 

<img width="500" alt="schermafbeelding 2015-09-10 om 14 08 29" src="https://cloud.githubusercontent.com/assets/14002770/9788101/12d0d144-57c8-11e5-83b8-04f7bd8ad272.png">

As you can see i made the airplane fly to RTM and then to STD. I would like the airplane to fly a straight line between the to fixes, yet the airplane overfly's the fix and then starts the turn, which causes the airplane being off-track. This only happens of course when the angle to turn is big enough to cause this problem. The way to fix this is to make a calculation based on airspeed (as turn radius is a function of airspeed) and angle the airplane has to turn, so the airplane comes on a straight line between the two fixes. I don't know if this is hard to fix, it is just my suggestion! 

3) Selecting all airplanes
The last thing i came across is when i type in a letter some airplanes will be selected including a picture of where they are heading and going. I would like to add a command which shows the path of all airplanes! We could use the word "all" to highlight all airplanes. 

<img width="600" alt="schermafbeelding 2015-09-10 om 14 09 53" src="https://cloud.githubusercontent.com/assets/14002770/9788268/b98b80ba-57c8-11e5-9488-a4a192af2da0.png">

As you can see, 3 airplanes are selected, but i would like to see a new function where you can select all airplanes!

To summarize:
- [ ] Make a new holding pattern
- [x] Change the calculation of turn to get the airplane fly a direct route between fixes
- [x] Make a command to select all airplanes

**Comment by glangford** [comment]
Good thoughts!

I don't have time to dig into these into the short term, but like your ideas. The "correct entry for holding" aspect is a bit messy as you know.  :)

(One suggestion for the future - in general, when proposing different enhancements it's good to create a separate issue for each. Then separate conversations about the advantages, disadvantages etc. can be tracked in one place, and when each is implemented the issue can be closed.)

**Comment by Maverick283** [comment]
1) I like the hold idea! There is a clear rule for hold entries, and as this is a simulator I think the fact that most planes will not fly a 1 minute outbound leg due to wind can be dropped. (How does wind affect flying aircraft anyways?). 

2) If the aircraft has an FMC, then it will most likely have the fix at the tip of its turn, still flying not a direct line from one fix to the other. If the aircraft is equipped with let's say 1 VOR receiver only, it will overfly the VOR, then make a turn to intercept the direct line and turn the other direction to follow the line once it is intercepted. Also, different pilots will fly different ways. I think, this could be a later enhancement to add a lot of reality.

3) I do like that. This might want to include STARs (once they are implemented), and similar "routes"...

**Comment by nielsvdweide** [comment]
@Maverick283 

1) What do you mean with most planes will not fly a 1 minute outbound leg due to wind? That's what a hold is all about. A hold consists of two turns and two legs of straight flying corrected for the wind. This will be 1 minute plus or minus 1 second for each knot of headwind or tailwind. So if you have 15 knots headwind you will fly 1 minute and 15 seconds on the inbound and then on the outbound you will fly for 45 seconds.

2) I've flown aircraft with only 1 VOR receiver, but most of the time you have got 2 instruments reading off  VOR stuff. If you have the CDI you can select the course so you can check whether you are on a specific radial or not. You also have got a RMI which is a bearing indicator which shows you an arrow which is pointing towards the beacon, thus showing the radial you're on. 

During my Instrument training i tracked a radial inbound to the station and depending on my ground speed i would turn before i pass the beacon. 2 miles before the beacon i would change the course to the next leg, maintaining my radial on the RMI. Then depending on my groundspeed i would take 10% of that, and that will be the distance i turn in front of the beacon. Most of the times works flawless. 

3) I think the last one is the best and needs to be implemented asap. I would also like a option where you can change the length of the path.

**Comment by Maverick283** [comment]
@nielsvdweide 

1) The hold is about flying a one minute inbound leg. No in order to achieve that with wind you will need a longer or shorter outbound leg (as wind is hardly ever 0 up there). But I believe this difference can be ignored for the simulation at this stage. (And the parenthesis question was for the programmers: How does wind affect airborn aircraft in the simulator)?

2) Well, that is good for you, all I'm saying here though is that not all pilots try to get a straight path between two fixes on a route, and I don't know of any rule that makes you do that. So, what I'm trying to say is that i don't think this is an issue with priority, but can be saved for later!

**Comment by nielsvdweide** [comment]
@Maverick283 

I agree on number 1 that wind can be ignored in holds. I don't know what wind does to aircraft in-game but i think that it doesn't really apply. I have never "observed" aircraft flying faster with a tailwind. I think it only applies for scoring but i'm not really sure.

2) I can understand! It's just the way our instructor told us! I don't think there is a way or rule about it, it's just a more commercial turn. It doesn't need priority, it's just something that will make it more realistic, i guess! 

I would like number 3 implemented soon :) What do you think?

**Comment by Maverick283** [comment]
@nielsvdweide  Well I like the idea of number 3 a lot, but I'm not the one programming here... I'm rather a person that just started to add ideas and I give suggestions base on my real world flying experience. Maybe you should start a new issue with number 3 only, and get rid of our discussion here (and either save the other two ideas for later or just make another 2 issues) so that the programmers can get to it ASAP and see what it is about.

**Comment by nielsvdweide** [comment]
@Maverick283 I made another issue for number 3. We can discuss over there about that. I've saved the other 2 items for a later moment. thank you!

**Comment by glangford** [comment]
I am just researching number 2 (routes between fixes) - assuming this is implemented, is it safe to assume for simulation purposes that every fix/waypoint can be a fly-by waypoint vs. flyover?

**Comment by nielsvdweide** [comment]
i think so, yes!

**Comment by glangford** [comment]
Here is a test:

![turn initiation](https://cloud.githubusercontent.com/assets/3035114/9983780/0a9b4582-5fd8-11e5-8b42-f84611f717a0.png)

**Comment by glangford** [comment]
> As you can see i made the airplane fly to RTM and then to STD. I would like the airplane to fly a straight line between the to fixes, yet the airplane overfly's the fix and then starts the turn, which causes the airplane being off-track.

Here is an image showing a similar RTM-STD turn with new physics; it's a 140 degree change in direction, which is started about 6.4km before RTM. Does that look about right? 

![rtm-std](https://cloud.githubusercontent.com/assets/3035114/9984125/3219ea8e-5fe0-11e5-91f4-c82b440f9df4.png)

**Comment by nielsvdweide** [comment]
Looking really good @glangford 😊

**Comment by nielsvdweide** [comment]
@glangford is this ready for use? :+1:

**Comment by glangford** [comment]
@nielsvdweide - not just yet, still some testing to do.

**Comment by nielsvdweide** [comment]
@glangford cool! Would love to use this :)

**Comment by glangford** [comment]
@nielsvdweide - while you are waiting, you can study 
http://www.ohio.edu/people/uijtdeha/ee6900_fms_00_overview.pdf

slides 48 and 49 for the upcoming quiz. :smiley:

**Comment by nielsvdweide** [comment]
@glangford haha no thank you! I've finished my atpl theory long ago :smile:

**Comment by nielsvdweide** [comment]
@glangford I want to try to make an airport but I have no experience with github at all! Where do you program and most important: where do you test it? I didn't find a place where i could contact you in private!

**Comment by zlsa** [comment]
@nielsvdweide, you can just clone the repository and run a local copy. If you're on Linux, you can use `python -m http.server` in the same directory as the source code (i.e. the directory with `index.html`), but I'm not sure how to do that on Windows/Mac.

You can just test the airport yourself locally, but the format has improved a lot recently thanks to @tedrek and @glangford's work, so the wiki page is out of date.

**Comment by glangford** [comment]
@nielsvdweide , in addition to @zlsa 's comment you can run on Mac (and I assume Windows) by simply  opening your local copy of index.html in the browser. To see examples of airports you might follow, see the format of KSFO.json or KJFK.json as examples. If you have any other questions, just post here.

**Comment by nielsvdweide** [comment]
I'm taking little steps but i'm getting there! Is there a small delay between updating files and showing it if i'm testing it? @glangford @zlsa

**Comment by glangford** [comment]
> Is there a small delay between updating files and showing it if i'm testing it?

The files won't reload automatically when you update them, so restarting the browser after updating is the way to go.

**Comment by tedrek** [comment]
I have ran into issues with the browser caching .json files.  I personally use chromium on linux and use a new incognito window to test changes to the .json files.  (I also have a local instance of nginx running as my web server).

Feel free to create a pull request asking for feedback on any aspect before you are ready for it to be merged.  This way we can have a discussion over code pieces in question by commenting directly on the code instead of in an issue where the code is separate.

**Comment by MaicomMR** [comment]
@tedrek I use Firefox on windows, the Google Chrome freeze on loading.

**Comment by nielsvdweide** [comment]
@glangford @tedrek I am good with testing my new airport at the moment. Is there a fast way to calculate where to put waypoints according to charts? (maybe even GPS coordinates or certain radial/dme?)

**Comment by glangford** [comment]
> Is there a fast way to calculate where to put waypoints according to charts?

I used opennav.com for most of the fixes/waypoints, and used the lat/long in the airport .json.

**Comment by nielsvdweide** [comment]
@Glangford opennav only provides me with coordinate data in minutes and seconds, for now converting those to decimals is fine! I've drawn the runway for now and a FAF which should be exactly in front of the runway, but it's just of the centerline. I will continue tomorrow!

For your interest; I'm making EDDN; Nurnberg

**Comment by zlsa** [comment]
You can open the settings for the developers console and check "Disable cache (while DevTools is open)". After that, keep it open and press Ctrl-F5 (or hold down the Reload button and select the Hard Reload option). Both of those have worked perfectly for me.

**Comment by glangford** [comment]
> opennav only provides me with coordinate data in minutes and seconds, for now converting those to decimals is fine!

Here is a trick to save time; as you are doing EDDN, if you look at the ERTES2T arrival for example, it has fix "DODAS". Rather than converting to decimals by hand, google "opennav dodas". You will see on the search results page that google has done the conversion, you can simply cut and paste.

**Comment by tedrek** [comment]
For coordinate data the code takes three formats:
- Decimal degrees - 'N47.112388112'
- Decimal minutes - 'N38d38.109808'
- Decimal seconds - 'N58d27m12.138'

You can also use the degree symbol instead of d, or a single quote instead of m or a space instead of either.  I did it this way precisely because opennav showed degree/minute/seconds but another source I was using was decimal degrees.

**Comment by erikquinn** [comment]
Item 1: Implemented in sim as of #443.
Item 2: Implemented in sim as of #183.
Item 3: Will not support this functionality.

Closing.
